database/query_builder.go

```diff
@@ -0,0 +1,254 @@
+}
+
+// Upsert returns an upsert statement for the query builders subject.
+func (qb *QueryBuilder) Upsert() (string, int) {
+	var updateColumns []string
+	if upserter, ok := qb.subject.(Upserter); ok {
+		updateColumns = qb.db.columnMap.Columns(upserter.Upsert())
+	} else {
+		updateColumns = qb.BuildColumns()
+	}
+
+	return qb.UpsertColumns(updateColumns...)
+}
+
+// UpsertColumns returns an upsert statement for the query builders subject and the specified update columns.
+func (qb *QueryBuilder) UpsertColumns(updateColumns ...string) (string, int) {
+	var clause, setFormat string
+	switch qb.db.DriverName() {
+	case MySQL:
+		clause = "ON DUPLICATE KEY UPDATE"
+		setFormat = `"%[1]s" = VALUES("%[1]s")`
+	case PostgreSQL:
+		clause = fmt.Sprintf("ON CONFLICT ON CONSTRAINT %s DO UPDATE SET", qb.getPgsqlOnConflictConstraint())
+		setFormat = `"%[1]s" = EXCLUDED."%[1]s"`
+	default:
+		panic("Driver unsupported: " + qb.db.DriverName())
+	}
+
+	set := make([]string, 0, len(updateColumns))
+	for _, col := range updateColumns {
+		set = append(set, fmt.Sprintf(setFormat, col))
+	}
+
+	insertColumns := qb.BuildColumns()
+
+	return fmt.Sprintf(
+		`INSERT INTO "%s" ("%s") VALUES (%s) %s %s`,
+		TableName(qb.subject),
+		strings.Join(insertColumns, `", "`),
+		fmt.Sprintf(":%s", strings.Join(insertColumns, ", :")),
+		clause,
+		strings.Join(set, ", "),
+	), len(insertColumns)
+}
+
+// Where returns a WHERE clause with named placeholder conditions built from the
+// specified scoper/column combined with the AND operator.
+func (qb *QueryBuilder) Where(subject any) (string, int) {
+	t := reflect.TypeOf(subject)
+	if t == nil { // Subject is a nil interface value.
+		return "", 0
+	}
+
+	var columns []string
+	if t.Kind() == reflect.String {
+		columns = []string{subject.(string)}
+	} else if t.Kind() == reflect.Struct || t.Kind() == reflect.Pointer {
+		if scoper, ok := subject.(Scoper); ok {
+			return qb.Where(scoper.Scope())
+		}
+
+		columns = qb.db.columnMap.Columns(subject)
+	} else { // This should never happen unless someone wants to do some silly things.
+		panic(fmt.Sprintf("qb.Where: unknown subject type provided: %q", t.Kind().String()))
+	}
+
+	where := make([]string, 0, len(columns))
+	for _, col := range columns {
+		where = append(where, fmt.Sprintf(`"%[1]s" = :%[1]s`, col))
+	}
+
+	return strings.Join(where, ` AND `), len(columns)
+}
+
+// BuildColumns returns all the Query Builder columns (if specified), otherwise they are
+// determined dynamically using its subject. Additionally, it checks whether columns need
+// to be excluded and proceeds accordingly.
+func (qb *QueryBuilder) BuildColumns() []string {
+	var columns []string
+	if len(qb.columns) > 0 {
+		columns = qb.columns
+	} else {
+		columns = qb.db.columnMap.Columns(qb.subject)
+	}
+
+	if len(qb.excludedColumns) > 0 {
+		columns = slices.DeleteFunc(append([]string(nil), columns...), func(column string) bool {
```

Comment:
> By doing so, the slices.DeleteFunc does not alter the referenced columns = qb.columns original slice. 

Actually, I wouldn't pay that much attention if this was the only rationale for this, but since `qb.columns` may contain columns mapped from the internal DB mapper, which explicitly warns you not to change the returned slice directly, I do it unconditionally.
https://github.com/Icinga/icinga-go-library/blob/2d47d95fe9d8d3d43d359de8569d9abdf6fecb9d/database/column_map.go#L12-L16